include/zephyr/instrumentation/instrumentation.h

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_INSTRUMENTATION_INSTRUMENTATION_H_
+#define ZEPHYR_INCLUDE_INSTRUMENTATION_INSTRUMENTATION_H_
+
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if !defined(__no_instrumentation__)
+#error "No toolchain support for __no_instrumentation__"
+#endif
+
+/**
+ * @brief Instrumentation event record types.
+ */
+enum __attribute__ ((__packed__)) instr_event_types {
+	INSTR_EVENT_META = 0,	/* Metadata record, defined in TLV format. Not used currently. */
+	INSTR_EVENT_ENTRY,	/* Callee entry event record, followed by instr_event. */
+	INSTR_EVENT_EXIT,	/* Callee exit event record, followed by instr_event. */
+	INSTR_EVENT_ENTRY_CTXT,	/* Entry event record, but with context */
+	INSTR_EVENT_EXIT_CTXT,	/* Exit event record, but with context */
+	INSTR_EVENT_PROFILE,	/* Profile events */
+	INSTR_EVENT_SCHED_IN,	/* Thread switched in scheduler event */
+	INSTR_EVENT_SCHED_OUT,	/* Thread switched out scheduler event */
+	INSTR_EVENT_NUM
+};
+
+/**
+ * @brief Header for the event records.
+ */
+struct instr_header {
+	/* Event type */
+	enum instr_event_types type;
+} __packed;
+
+struct instr_event_context {
+	/** Arch-specific mode indicator (thread mode, interrupt mode, etc.). */
+	uint8_t mode : 3;
+	/** CPU number. */
+	uint8_t cpu : 3;
+	/** Reserved for future use. */
+	uint8_t rsvd : 2;
+	/** Thread ID (correlate values with thread lookup table). */
+	void* thread_id;
+	/* Thread name (that can be compacted with the correlate lookup table). */
+	char thread_name[20];
```

Comment:
this should be guarded by `CONFIG_THREAD_NAME`

Reply:
Done.